### PR TITLE
Fix propagating post-save mutations

### PR DESCRIPTION
**Pull request by booleangate**
This addresses the bug I reported in https://github.com/ent/ent/issues/2523.  I've broken commits up so you can review just the first two in isolation (a [failing test](https://github.com/ent/ent/pull/2524/commits/6c325d62fdca79ff20c8eda2b77e9c6a928e10cf), and the [fix](https://github.com/ent/ent/pull/2524/commits/49930531319ad1023114698a7fdf6f1644196594)) to cut out all the noise from the regeneration.

📓  I had to branch my changes off of the fix/bump-atlas branch in order to be able to run `go generate ./...`.  I'm just waiting for that to merge, then I'll rebase.

### CODE REVIEW

File: entc/gen/template/builder/create.tmpl [hunk 1]
```diff
@@ -83,8 +83,10 @@ func ({{ $receiver }} *{{ $builder }}) Save(ctx context.Context) (*{{ $.Name }},
 			}
 			mut = {{ $receiver }}.hooks[i](mut)
 		}
-		if _, err := mut.Mutate(ctx, {{ $mutation }}); err != nil {
+		if v, err := mut.Mutate(ctx, {{ $mutation }}); err != nil {
 			return nil, err
+		} else if v, ok := v.(*{{ $.Name }}); ok {
+			node = v
```
Comment (booleangate) [inline entc/gen/template/builder/create.tmpl:89]: 📓  The fix

File: entc/integration/hooks/hooks_test.go [hunk 2]
```diff
@@ -212,6 +212,32 @@ func TestPostCreation(t *testing.T) {
 	client.Card.CreateBulk(client.Card.Create().SetNumber("12345")).SaveX(ctx)
 }
 
+func TestUpdateAfterCreation(t *testing.T) {
```
Comment (booleangate) [inline entc/integration/hooks/hooks_test.go:215]: 📓  This test expresses the bug.

File: entc/gen/template/builder/create.tmpl [hunk 3]
```diff
@@ -83,8 +83,10 @@ func ({{ $receiver }} *{{ $builder }}) Save(ctx context.Context) (*{{ $.Name }},
...
 		}
```
Comment (a8m) [inline entc/gen/template/builder/create.tmpl:90]:
```suggestion
		v, err := mut.Mutate(ctx, {{ $mutation }})
		if err != nil {
			return nil, err
		}
		 nv, ok := v.(*{{ $.Name }})
		 if !ok {
		 	 return nil, fmt.Errorf("unexpected node type %T returned from {{ $.MutationName }}", v)
		 }
		 node = nv
```

File: entc/integration/hooks/hooks_test.go [hunk 4]
```diff
@@ -212,6 +212,32 @@ func TestPostCreation(t *testing.T) {
...
+			// Important to return a completely new model; don't just mutate the existing user.
+			newUser := m.Client().User.UpdateOneID(existingUser.ID).
+				SetVersion(2).
+				SaveX(ctx)
+			return newUser, nil
+		})
+	}, ent.OpCreate))
+
+	u := client.User.Create().SetName("a8m").SetVersion(1).SaveX(ctx)
+	require.Equal(t, u.Version, 2, "version mutation in hook should have propagated back to call site")
```
Comment (a8m) [inline entc/integration/hooks/hooks_test.go:238]: 💯

File: entc/integration/hooks/hooks_test.go [hunk 5]
```diff
@@ -212,6 +212,32 @@ func TestPostCreation(t *testing.T) {
...
+	client := enttest.Open(t, "sqlite3", "file:ent?mode=memory&cache=shared&_fk=1")
+	defer client.Close()
+	client.User.Use(hook.On(func(next ent.Mutator) ent.Mutator {
+		return hook.UserFunc(func(ctx context.Context, m *ent.UserMutation) (ent.Value, error) {
+			value, err := next.Mutate(ctx, m)
+			if err != nil {
+				return nil, err
+			}
+			existingUser, ok := value.(*ent.User)
+			require.True(t, ok, "value should be of type %T", existingUser)
```
Comment (a8m) [inline entc/integration/hooks/hooks_test.go:226]:
```suggestion
			require.Truef(t, ok, "value should be of type %T", existingUser)
```

File: entc/integration/hooks/hooks_test.go [hunk 6]
```diff
@@ -212,6 +212,32 @@ func TestPostCreation(t *testing.T) {
...
+			require.Equal(t, existingUser.Version, 1, "version does not match the original value")
+
+			// Important to return a completely new model; don't just mutate the existing user.
```
Comment (a8m) [inline entc/integration/hooks/hooks_test.go:229]:
```suggestion
			// After the user was created, return its updated version (a new object).
```

**Review by a8m — COMMENTED** [review]
Thanks for the contribution @booleangate! The changes look good. 

Please, fix the minor suggestions and I'll merge this PR. Also note, that the changes need to be applied on the other hook types as well (update, update-one, etc). Would you like to contribute them in different PRs?

**Comment by booleangate** [comment]
> Also note, that the changes need to be applied on the other hook types as well (update, update-one, etc). Would you like to contribute them in different PRs?

@a8m I'm afraid I don't follow but happy to learn.  The signature on the update and delete clients is `Save(...) (int, error)` and `Exec(...) (int, error)` respectively.  Because they return affected row count and not the model itself, I don't see how we could make a similar change to those clients.

**Comment by a8m** [comment]
> > Also note, that the changes need to be applied on the other hook types as well (update, update-one, etc). Would you like to contribute them in different PRs?
> 
> @a8m I'm afraid I don't follow but I'm happy to learn. The signature on the update and delete clients is `Save(...) (int, error)` and `Exec(...) (int, error)` respectively. Because they return affected row count and not the model itself, I don't see how we could make a similar change to those clients.

Let's start with update-one - https://github.com/ent/ent/blob/master/entc/gen/template/builder/update.tmpl#L200-L202

**Comment by a8m** [comment]
CI is red. Please, run `go generate ./...` from the root of the project. Thanks 🙏

**Comment by a8m** [comment]
@booleangate can you rebase your branch on master and create a new PR?

**Comment by booleangate** [comment]
@a8m yup, working on a new PR now with the UpdateOne changes.

**Pull request closed without merging**